### DISCUSSION TITLE
Architecture Research: Recurrent-Depth Transformer (RDT) — component roadmap

### DISCUSSION THREAD

**Discussion by kolkov** [born-ml/born#60] [Ideas]
## Context

We've been tracking recent ML architecture innovations relevant to Born's roadmap. One development worth discussing is the **Recurrent-Depth Transformer (RDT)** architecture, recently reconstructed as an open-source PyTorch implementation ([OpenMythos](https://github.com/kyegomez/OpenMythos)).

This post summarizes our research findings and lays out how RDT components map to Born's existing roadmap. We'd love community input on priorities.

## What is a Recurrent-Depth Transformer?

Standard transformers stack N unique layers sequentially. An RDT instead uses a **fixed set of weights applied iteratively in a loop** (up to T=16 steps per forward pass):

```
Input → Embedding
  → Prelude (standard transformer layers, run once)
    → Recurrent Block (loop T iterations):
        h_{t+1} = A·h_t + B·e + Transformer(h_t, e)
    → Coda (standard transformer layers, run once)
  → Output head
```

The key claim: **770M parameters can match a 1.3B standard transformer** on identical training data (~2x parameter efficiency). The model trades compute (more iterations) for parameters (fewer weights).

## Architecture Components

| Component | What it does | Origin |
|-----------|-------------|--------|
| **Recurrent Loop** | Iterates transformer block T times with shared weights | RDT (Parcae paper) |
| **LTI Injection** | `A·h + B·e` — stable linear state update per iteration | RDT |
| **MoE FFN** | Sparse top-K expert routing replaces dense FFN | DeepSeek-V3 |
| **Multi-Latent Attention** | Compressed KV cache via low-rank projection (10-20x savings) | DeepSeek-V2 |
| **Depth-Wise LoRA** | Per-iteration adapters for behavioral differentiation | RDT + LoRA |
| **ACT Halting** | Dynamic early termination — easy tokens use fewer iterations | Graves (2016) |

## What Born already has vs what's needed

| Component | Born status | Effort |
|-----------|------------|--------|
| Transformer layers (Linear, RMSNorm, Attention, MLP) | ✅ Supported | — |
| Embedding, Softmax, MatMul | ✅ Supported | — |
| LTI Injection (A·h + B·e) | ✅ Just MatMul + Add | — |
| ONNX/GGUF import | ✅ Supported | — |
| **Multi-Latent Attention** | ❌ Not yet | Medium |
| **MoE Routing** (top-K sparse) | ❌ Not yet | High |
| **Recurrent Loop** primitive | ❌ Not yet | Medium |
| **Depth-Wise LoRA** | ❌ Not yet | Medium |
| **ACT Halting** | ❌ Not yet | Medium |

## Our assessment

**RDT is interesting but unproven.** Key considerations:

1. **No trained RDT models exist yet.** OpenMythos is code only — no weights, no GGUF/ONNX exports. Born supports both training and inference, so we could potentially train RDT models ourselves once the primitives are in place — but the architecture needs independent validation first.

2. **The efficiency claim is not independently reproduced.** The "770M = 1.3B" result comes from a single paper. No public benchmarks.

3. **RDT trades latency for parameters.** Each forward pass runs T loop iterations → slower inference. However, the ~2x parameter efficiency means cheaper training and smaller model footprint for deployment.

4. **Several RDT components are independently valuable — for both training and inference:**
   - **Multi-Latent Attention** — needed for DeepSeek V2/V3/V4 (10-20x KV cache reduction)
   - **MoE Routing** — needed for DeepSeek V3/V4, Mixtral, and other production MoE models
   - **LoRA adapters** — useful for domain adaptation, fine-tuning, and personalization

## Proposed timeline

| Component | Target | Rationale |
|-----------|--------|-----------|
| Multi-Latent Attention | Near-term | High value — enables DeepSeek V2/V3/V4 |
| MoE Routing | Near-term | High value — enables DeepSeek V4, Mixtral |
| LoRA adapters | Mid-term | Medium value — domain adaptation, complements Engram tables |
| Recurrent Loop + ACT | When ready | Enables RDT training and inference — needs validation of efficiency claims first |

## Questions for the community

1. **Which component would be most useful to you?** Multi-Latent Attention (memory savings), MoE (model support), or LoRA (adaptation)?

2. **Are you tracking any other RDT-like architectures** (looped transformers, deep equilibrium models, neural ODEs) that we should be aware of?

3. **Would you be interested in contributing** to any of these components? MoE routing and Multi-Latent Attention are both well-scoped tasks with clear reference implementations in DeepSeek.

## References

- [OpenMythos (GitHub)](https://github.com/kyegomez/OpenMythos) — PyTorch RDT implementation
- [MarkTechPost coverage (2026-04-19)](https://www.marktechpost.com/2026/04/19/meet-openmythos-an-open-source-pytorch-reconstruction-of-claude-mythos-where-770m-parameters-match-a-1-3b-transformer/)
- Parcae (Prairie et al., 2026) — original RDT hypothesis paper
- [DeepSeek-V2 paper](https://arxiv.org/abs/2405.04434) — Multi-Latent Attention
- [DeepSeekMoE paper](https://arxiv.org/abs/2401.06066) — Fine-grained Mixture-of-Experts
